### PR TITLE
profiles layer: Set instance version from profile

### LINKED ISSUE #118: Allow overriding the instance application apiVersion when SIMULATE_API_VERSION_BIT is set
Currently, this can't be done because the Profiles file is not loaded when calling CreateInstance but this should be resolve applying Profiles with higher Vulkan version requirements than the Vulkan application.

This is an item for SDK 1.3.204.1

### PR DESCRIPTION
**Pull request by ziga-lunarg**
Closes #118

### CODE REVIEW

File: layer/profiles.cpp [hunk 1]
```diff
@@ -7553,55 +7607,95 @@ VKAPI_ATTR VkResult VKAPI_CALL CreateInstance(const VkInstanceCreateInfo *pCreat
...
+    bool changed_version = false;
+    uint32_t profile_api_version = json_loader.GetProfileApiVersion();
+    if (VK_VERSION_MAJOR(requested_version) < VK_VERSION_MAJOR(profile_api_version) ||
+        VK_VERSION_MINOR(requested_version) < VK_VERSION_MINOR(profile_api_version)) {
+        LogMessage(DEBUG_REPORT_WARNING_BIT,
+                   ::format("The Profiles Layer requires API version %s, but instance is being created "
+                            "with API version %s. The API version will be set to %s.\n",
+                            StringAPIVersion(requested_version).c_str(), StringAPIVersion(profile_api_version).c_str(),
+                            StringAPIVersion(requested_version).c_str()));
+        requested_version = profile_api_version;
```
Comment (ziga-lunarg) [inline layer/profiles.cpp]: Should this be under a setting? `SIMULATE_API_VERSION_BIT` or a new one?

File: layer/profiles.cpp [hunk 2]
```diff
@@ -4680,40 +4715,40 @@ VkResult JsonLoader::ReadProfile(const Json::Value root, const std::vector<std::
                 AddPromotedExtensions(properties_api_version);
             }
         } else if (layer_settings->simulate_capabilities & SIMULATE_API_VERSION_BIT) {
-            AddPromotedExtensions(this->profile_api_version);
+            AddPromotedExtensions(this->profile_api_version_);
         }
 
-        if (VK_VERSION_PATCH(this->profile_api_version) > VK_VERSION_PATCH(pdd_.physical_device_properties_.apiVersion)) {
+        if (VK_VERSION_PATCH(this->profile_api_version_) > VK_VERSION_PATCH(pdd_->physical_device_properties_.apiVersion)) {
```
Comment (aqnuep) [inline layer/profiles.cpp]: `VK_VERSION_PATCH` (actually deprecated, and `VK_API_VERSION_PATCH` should be used instead) is only returning the patch version from the full version. Are you sure that's what you meant to use here and not compare the major+minor version instead?

File: layer/profiles.cpp [hunk 3]
```diff
@@ -4954,39 +4990,66 @@ VkResult JsonLoader::LoadFile(std::string filename) {
...
+
+    /*if (VK_VERSION_MINOR(this->profile_api_version_) > VK_VERSION_MINOR(requested_version)) {
+        LogMessage(DEBUG_REPORT_WARNING_BIT,
+                   format("The application requested a Vulkan 1.%d instance however the selected profile requires Vulkan 1.%d. "
+                          "Some profile capabilities might not be applied, we recommend changing the Vulkan application to request "
+                          "a Vulkan 1.%d instance.\n",
+                          VK_VERSION_MINOR(requested_version), VK_VERSION_MINOR(this->profile_api_version_),
+                          VK_VERSION_MINOR(this->profile_api_version_))
+                       .c_str());
+    }*/
```
Comment (aqnuep) [inline layer/profiles.cpp]: I guess this commented out code should be deleted.

**Review by aqnuep — COMMENTED** [review]
I can't tell whether this behaves as expected, as the code interactions are non-trivial, so it would be great to have some sort of test for this.

As discussed, we'd probably need some sort of Vulkan API mock to be able to catch what gets passed to `vkCreateInstance`, although it's not trivial how to do it in a setting like this, where we'd actually have to capture `vkCreateInstance` within the layer chain.

Thus, personally, I'd be happy enough if this is verified manually for now if an automated test seems unfeasible.

**Review by christophe-lunarg — APPROVED** [review]
I made some fixes:
- Improved messages
- Protected application API version override with SIMULATE_API_VERSION_BIT
- Fixed usage of deprecated Vulkan macros
- When a profile name is not found, it's just a warning and we load the default profile

**Pull request merged**